### PR TITLE
Add decimal value support

### LINKED ISSUE #265: AMQP decimal data type support missed
Also, we need to check overall AMQP data types support and correctness.

### PR DESCRIPTION
**Pull request by pinepain**
This PR add decimals decoding support. Closes #265.

### CODE REVIEW

**Comment by lstrojny** [comment]
Thanks for that. Don't we rather want to encode decimals as string (as e.g. bcmath does)?

**Comment by lstrojny** [comment]
See my comment in #265

**Comment by lstrojny** [comment]
At @InterNations we found the need for proper array and timestamp support so that we can use RabbitMQs dead lettering support. I've opened an issue and started working on it. @pinepain can you give me feedback on the plan for #271?

**Comment by pinepain** [comment]
@lstrojny Please, take a look, I added decimals support and cleanup a bit.

File: amqp_basic_properties.c [hunk 1]
```diff
@@ -398,7 +399,11 @@ void parse_amqp_table(amqp_table_t *table, zval *result TSRMLS_DC) {
                 ZVAL_LONG(PHP5to7_MAYBE_PTR(value), entry->value.value.i64);
                 break;
             case AMQP_FIELD_KIND_U64:
-                ZVAL_LONG(PHP5to7_MAYBE_PTR(value), entry->value.value.i64);
+                if (entry->value.value.u64 > LONG_MAX) {
+                    ZVAL_DOUBLE(PHP5to7_MAYBE_PTR(value), entry->value.value.u64);
```
Comment (lstrojny) [inline amqp_basic_properties.c]: We should probably introduce an 64 bit integer object as well and encode it as string. But that’s out of scope for this PR.

File: stubs/AMQPDecimal.php [hunk 2]
```diff
@@ -0,0 +1,30 @@
+<?php
+
+/**
+ * stub class representing AMQPDecimal from pecl-amqp
+ */
+final class AMQPDecimal
+{
+    const E_MIN = 0;
```
Comment (lstrojny) [inline stubs/AMQPDecimal.php]: Wanna call this EXPONENT_MIN and EXPONENT_MAX?

File: stubs/AMQPDecimal.php [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+    const E_MAX = 255;
+    const N_MIN = 0;
```
Comment (lstrojny) [inline stubs/AMQPDecimal.php]: Wanna call this significand or mantissa and therefore SIGNIFICAND_MAX/MIN or MANTISSA_MAX/MIN?

**Comment by lstrojny** [comment]
Looks good!

File: amqp_basic_properties.c (hunk 1 above)
Comment (pinepain) [inline amqp_basic_properties.c]: Indeed. This is just a quick fix. Also, messages sent and received with php-amqp was not affected with this mistake even before a fix.

File: stubs/AMQPDecimal.php (hunk 2 above)
Comment (pinepain) [inline stubs/AMQPDecimal.php]: Yeah, will do. I think renaming `getE` to `getExponent` and `getN` to `getSignificand` will also a good thing.

File: amqp_decimal.c [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+
+PHP_MINIT_FUNCTION(amqp_decimal)
+{
+	zend_class_entry ce;
+
+	INIT_CLASS_ENTRY(ce, "AMQPDecimal", amqp_decimal_class_functions);
+	this_ce = zend_register_internal_class(&ce TSRMLS_CC);
+	this_ce->ce_flags = this_ce->ce_flags | PHP5to7_ZEND_ACC_FINAL_CLASS;
+
+    zend_declare_class_constant_long(this_ce, ZEND_STRL("EXPONENT_MIN"), AMQP_DECIMAL_E_MIN TSRMLS_CC);
```
Comment (lstrojny) [inline amqp_decimal.c]: Could you rename the C variable names as well? I feel that would make it easier to follow the code

**Comment by lstrojny** [comment]
Only one remark from my side. Afterwards I would merge and release a new version. Do we go straight to 1.9.0?

File: amqp_decimal.c (hunk 4 above)
Comment (pinepain) [inline amqp_decimal.c]: Done.

**Comment by pinepain** [comment]
Yeah, 1.9.0 should be ok, but I'd rather go with beta first.

**Comment by lstrojny** [comment]
Sure

**Pull request merged**